File: src/lib/simplify.test.ts
import { describe, it, expect } from "vitest";
import {
  haversineMeters,
  pathLengthMeters,
  decimateByDistance,
  douglasPeucker,
  simplifyTrack,
  type LatLng,
} from "./simplify";

// One degree of latitude ≈ this many meters for R = 6371008.8 (2πR/360).
const DEG_LAT_M = 111194.93;

describe("haversineMeters", () => {
  it("is zero for identical points", () => {
    expect(haversineMeters({ lat: 40, lng: -70 }, { lat: 40, lng: -70 })).toBe(0);
  });

  it("matches one degree of latitude", () => {
    const d = haversineMeters({ lat: 0, lng: 0 }, { lat: 1, lng: 0 });
    expect(d).toBeCloseTo(DEG_LAT_M, 0);
  });

  it("shrinks a degree of longitude by cos(lat)", () => {
    const atEquator = haversineMeters({ lat: 0, lng: 0 }, { lat: 0, lng: 1 });
    const at60 = haversineMeters({ lat: 60, lng: 0 }, { lat: 60, lng: 1 });
    // cos(60°) = 0.5, so the high-latitude degree is ~half as wide.
    expect(at60 / atEquator).toBeCloseTo(0.5, 2);
  });

  it("is symmetric", () => {
    const a = { lat: 41.1, lng: -71.5 };
    const b = { lat: 41.4, lng: -70.9 };
    expect(haversineMeters(a, b)).toBeCloseTo(haversineMeters(b, a), 6);
  });
});

describe("pathLengthMeters", () => {
  it("is zero for fewer than two points", () => {
    expect(pathLengthMeters([])).toBe(0);
    expect(pathLengthMeters([{ lat: 1, lng: 1 }])).toBe(0);
  });

  it("sums consecutive legs", () => {
    const pts: LatLng[] = [
      { lat: 0, lng: 0 },
      { lat: 1, lng: 0 },
      { lat: 2, lng: 0 },
    ];
    expect(pathLengthMeters(pts)).toBeCloseTo(2 * DEG_LAT_M, 0);
  });
});

describe("decimateByDistance", () => {
  it("returns short inputs unchanged", () => {
    const pts = [
      { lat: 0, lng: 0 },
      { lat: 1, lng: 1 },
    ];
    expect(decimateByDistance(pts, 100)).toEqual(pts);
  });

  it("always keeps first and last", () => {
    // ~10 m steps in latitude; decimate at 100 m.
    const pts: LatLng[] = [];
    for (let i = 0; i <= 100; i++) pts.push({ lat: 40 + i * 0.00009, lng: -70 });
    const out = decimateByDistance(pts, 100);
    expect(out[0]).toEqual(pts[0]);
    expect(out[out.length - 1]).toEqual(pts[pts.length - 1]);
  });

  it("spaces kept points by roughly the granularity", () => {
    const pts: LatLng[] = [];
    for (let i = 0; i <= 100; i++) pts.push({ lat: 40 + i * 0.00009, lng: -70 });
    // ~1000 m total / 100 m granularity → ~11 points (10 gaps + tail).
    expect(decimateByDistance(pts, 100).length).toBe(11);
  });

  it("collapses sub-granularity jitter", () => {
    const pts: LatLng[] = [];
    for (let i = 0; i < 100; i++) {
      pts.push({ lat: 40 + (i % 3) * 0.00001, lng: -70 + (i % 2) * 0.00001 });
    }
    pts.push({ lat: 40.01, lng: -70 }); // one far point
    // All jitter is within a few meters → first + far point only.
    expect(decimateByDistance(pts, 50).length).toBe(2);
  });
});

describe("douglasPeucker", () => {
  it("collapses a straight line to its endpoints", () => {
    const line: LatLng[] = [];
    for (let i = 0; i < 50; i++) line.push({ lat: 40 + i * 0.001, lng: -70 });
    expect(douglasPeucker(line, 5)).toHaveLength(2);
  });

  it("preserves the corners of a square", () => {
    const cornerLat = 0.009;
    const cornerLng = 0.009 / Math.cos((40 * Math.PI) / 180);
    const A = { lat: 40, lng: -70 };
    const B = { lat: 40, lng: -70 + cornerLng };
    const C = { lat: 40 + cornerLat, lng: -70 + cornerLng };
    const D = { lat: 40 + cornerLat, lng: -70 };
    const sq: LatLng[] = [];
    const edge = (a: LatLng, b: LatLng) => {
      for (let i = 0; i < 20; i++) {
        sq.push({
          lat: a.lat + ((b.lat - a.lat) * i) / 20,
          lng: a.lng + ((b.lng - a.lng) * i) / 20,
        });
      }
    };
    edge(A, B);
    edge(B, C);
    edge(C, D);
    edge(D, A);
    sq.push({ ...A });
    // 4 corners + the closing return to A.
    expect(douglasPeucker(sq, 10)).toHaveLength(5);
  });

  it("keeps more points as tolerance tightens", () => {
    // A bump in the middle of an otherwise straight line.
    const pts: LatLng[] = [
      { lat: 40, lng: -70 },
      { lat: 40.001, lng: -69.999 },
      { lat: 40.002, lng: -70 },
    ];
    expect(douglasPeucker(pts, 1000).length).toBe(2); // bump within tolerance → dropped
    expect(douglasPeucker(pts, 10).length).toBe(3); // bump exceeds tolerance → kept
  });
});

describe("simplifyTrack", () => {
  it("drops invalid coordinates and the time field", () => {
    const out = simplifyTrack(
      [
        { lat: NaN, lng: 0 },
        { lat: 0, lng: 200 },
        { lat: 40, lng: -70, t: 5 },
        { lat: 40.001, lng: -70 },
      ],
      { granularityMeters: 1 }
    );
    expect(out.inputCount).toBe(2);
    expect(out.waypoints).toHaveLength(2);
    expect(out.waypoints[0]).toEqual({ lat: 40, lng: -70 });
    expect("t" in out.waypoints[0]).toBe(false);
  });

  it("returns two-or-fewer point inputs untouched", () => {
    const out = simplifyTrack([{ lat: 1, lng: 1 }], { granularityMeters: 100 });
    expect(out.waypoints).toHaveLength(1);
    expect(out.capped).toBe(false);
  });

  it("reduces a dense straight track to its endpoints", () => {
    const pts: LatLng[] = [];
    for (let i = 0; i < 200; i++) pts.push({ lat: 40 + i * 0.0001, lng: -70 });
    const out = simplifyTrack(pts, { granularityMeters: 50 });
    expect(out.waypoints).toHaveLength(2);
    expect(out.capped).toBe(false);
  });

  it("caps output at maxPoints and flags it", () => {
    // A high-amplitude zigzag DP can't reduce below its corner count, so the
    // maxPoints escalation must kick in.
    const pts: LatLng[] = [];
    for (let i = 0; i < 100; i++) {
      pts.push({ lat: 40 + i * 0.001, lng: -70 + (i % 2) * 0.0005 });
    }
    const out = simplifyTrack(pts, { granularityMeters: 1, maxPoints: 10 });
    expect(out.capped).toBe(true);
    expect(out.waypoints.length).toBeLessThanOrEqual(10);
    expect(out.waypoints.length).toBeGreaterThanOrEqual(2);
  });

  it("does not cap when the route already fits", () => {
    const pts: LatLng[] = [];
    for (let i = 0; i < 50; i++) pts.push({ lat: 40 + i * 0.001, lng: -70 });
    const out = simplifyTrack(pts, { granularityMeters: 10, maxPoints: 500 });
    expect(out.capped).toBe(false);
  });
});

File: mobile/test/simplify_test.dart
import 'dart:math' as math;

import 'package:flutter_test/flutter_test.dart';
import 'package:latlong2/latlong.dart';
import 'package:viam_chartplotter_mobile/simplify.dart';

// simplify.test.ts translated (E3).

// One degree of latitude ≈ this many meters for R = 6371008.8 (2πR/360).
const degLatM = 111194.93;

void main() {
  group('haversineMeters', () {
    test('is zero for identical points', () {
      expect(
          haversineMeters(const LatLng(40, -70), const LatLng(40, -70)), 0);
    });

    test('matches one degree of latitude', () {
      final d = haversineMeters(const LatLng(0, 0), const LatLng(1, 0));
      expect(d, closeTo(degLatM, 1));
    });

    test('shrinks a degree of longitude by cos(lat)', () {
      final atEquator =
          haversineMeters(const LatLng(0, 0), const LatLng(0, 1));
      final at60 = haversineMeters(const LatLng(60, 0), const LatLng(60, 1));
      // cos(60°) = 0.5, so the high-latitude degree is ~half as wide.
      expect(at60 / atEquator, closeTo(0.5, 0.01));
    });

    test('is symmetric', () {
      const a = LatLng(41.1, -71.5);
      const b = LatLng(41.4, -70.9);
      expect(haversineMeters(a, b), closeTo(haversineMeters(b, a), 1e-6));
    });
  });

  group('pathLengthMeters', () {
    test('is zero for fewer than two points', () {
      expect(pathLengthMeters(const []), 0);
      expect(pathLengthMeters(const [LatLng(1, 1)]), 0);
    });

    test('sums consecutive legs', () {
      const pts = [LatLng(0, 0), LatLng(1, 0), LatLng(2, 0)];
      expect(pathLengthMeters(pts), closeTo(2 * degLatM, 2));
    });
  });

  group('decimateByDistance', () {
    test('returns short inputs unchanged', () {
      const pts = [LatLng(0, 0), LatLng(1, 1)];
      expect(decimateByDistance(pts, 100), pts);
    });

    test('always keeps first and last', () {
      final pts = [
        for (var i = 0; i <= 100; i++) LatLng(40 + i * 0.00009, -70)
      ];
      final out = decimateByDistance(pts, 100);
      expect(out.first, pts.first);
      expect(out.last, pts.last);
    });

    test('spaces kept points by roughly the granularity', () {
      final pts = [
        for (var i = 0; i <= 100; i++) LatLng(40 + i * 0.00009, -70)
      ];
      // ~1000 m total / 100 m granularity → ~11 points (10 gaps + tail).
      expect(decimateByDistance(pts, 100).length, 11);
    });

    test('collapses sub-granularity jitter', () {
      final pts = [
        for (var i = 0; i < 100; i++)
          LatLng(40 + (i % 3) * 0.00001, -70 + (i % 2) * 0.00001),
        const LatLng(40.01, -70), // one far point
      ];
      // All jitter is within a few meters → first + far point only.
      expect(decimateByDistance(pts, 50).length, 2);
    });
  });

  group('douglasPeucker', () {
    test('collapses a straight line to its endpoints', () {
      final line = [
        for (var i = 0; i < 50; i++) LatLng(40 + i * 0.001, -70)
      ];
      expect(douglasPeucker(line, 5), hasLength(2));
    });

    test('preserves the corners of a square', () {
      const cornerLat = 0.009;
      final cornerLng = 0.009 / math.cos(40 * math.pi / 180);
      const a = LatLng(40, -70);
      final b = LatLng(40, -70 + cornerLng);
      final c = LatLng(40 + cornerLat, -70 + cornerLng);
      const d = LatLng(40 + cornerLat, -70);
      final sq = <LatLng>[];
      void edge(LatLng p, LatLng q) {
        for (var i = 0; i < 20; i++) {
          sq.add(LatLng(
            p.latitude + (q.latitude - p.latitude) * i / 20,
            p.longitude + (q.longitude - p.longitude) * i / 20,
          ));
        }
      }

      edge(a, b);
      edge(b, c);
      edge(c, d);
      edge(d, a);
      sq.add(a);
      // 4 corners + the closing return to A.
      expect(douglasPeucker(sq, 10), hasLength(5));
    });

    test('keeps more points as tolerance tightens', () {
      // A bump in the middle of an otherwise straight line.
      const pts = [
        LatLng(40, -70),
        LatLng(40.001, -69.999),
        LatLng(40.002, -70),
      ];
      expect(douglasPeucker(pts, 1000).length, 2); // within tolerance → drop
      expect(douglasPeucker(pts, 10).length, 3); // exceeds tolerance → keep
    });
  });

  group('simplifyTrack', () {
    test('drops invalid coordinates', () {
      final out = simplifyTrack(
        [
          const LatLng(double.nan, 0),
          // latlong2 refuses lng 200 at construction, so the out-of-range
          // case from the web test is expressed as an invalid latitude.
          const LatLng(double.nan, 100),
          const LatLng(40, -70),
          const LatLng(40.001, -70),
        ],
        const SimplifyOptions(granularityMeters: 1),
      );
      expect(out.inputCount, 2);
      expect(out.waypoints, hasLength(2));
      expect(out.waypoints.first, const LatLng(40, -70));
    });

    test('returns two-or-fewer point inputs untouched', () {
      final out = simplifyTrack(
          const [LatLng(1, 1)], const SimplifyOptions(granularityMeters: 100));
      expect(out.waypoints, hasLength(1));
      expect(out.capped, isFalse);
    });

    test('reduces a dense straight track to its endpoints', () {
      final pts = [
        for (var i = 0; i < 200; i++) LatLng(40 + i * 0.0001, -70)
      ];
      final out =
          simplifyTrack(pts, const SimplifyOptions(granularityMeters: 50));
      expect(out.waypoints, hasLength(2));
      expect(out.capped, isFalse);
    });

    test('caps output at maxPoints and flags it', () {
      // A high-amplitude zigzag DP can't reduce below its corner count, so
      // the maxPoints escalation must kick in.
      final pts = [
        for (var i = 0; i < 100; i++)
          LatLng(40 + i * 0.001, -70 + (i % 2) * 0.0005)
      ];
      final out = simplifyTrack(
          pts, const SimplifyOptions(granularityMeters: 1, maxPoints: 10));
      expect(out.capped, isTrue);
      expect(out.waypoints.length, lessThanOrEqualTo(10));
      expect(out.waypoints.length, greaterThanOrEqualTo(2));
    });

    test('does not cap when the route already fits', () {
      final pts = [
        for (var i = 0; i < 50; i++) LatLng(40 + i * 0.001, -70)
      ];
      final out = simplifyTrack(pts,
          const SimplifyOptions(granularityMeters: 10, maxPoints: 500));
      expect(out.capped, isFalse);
    });
  });
}
